src/app/data-planes/data/index.ts

```diff
@@ -0,0 +1,237 @@
+// generated protobuf or openapi imports
+import { DataplaneNetworking } from './networking'
+import { Dataplane as GeneratedDataplane } from '@/data/proto/mesh/v1alpha1/dataplane'
+import {
+  DataplaneInsight as GeneratedDataplaneInsight,
+  DiscoverySubscription as GeneratedDiscoverySubscription,
+  Version as GeneratedVersion,
+  KumaDpVersion as GeneratedKumaDpVersion,
+  EnvoyVersion as GeneratedEnvoyVersion,
+} from '@/data/proto/mesh/v1alpha1/dataplane_insight'
+import { DataplaneOverview as GeneratedDataplaneOverview } from '@/data/proto/mesh/v1alpha1/dataplane_overview'
+// end generated protobuf or openapi imports
+import type {
+  PaginatedApiListResponse as CollectionResponse,
+} from '@/types/api.d'
+
+type Label = {
+  label: string
+  value: string
+}
+type Dependencies = {
+  envoy?: GeneratedEnvoyVersion
+  kumaDp?: GeneratedKumaDpVersion,
+} & Record<string, {version: string}>
+
+export interface Entity {
+  type: string
+  name: string
+  creationTime: string
+  modificationTime: string
+}
+export const Entity = {
+  fromJSON: (object: any): Entity => {
+    return {
+      name: object.name || '',
+      type: object.type || '',
+      creationTime: object.creationTime || '',
+      modificationTime: object.modificationTime || '',
+    }
+  },
+}
+
+export type Dataplane = GeneratedDataplane & {
```

Comment:
I'm not sure how I feel about the GeneratedDataplane being accessible on the end object. Isn't one of the point of this DTO to abstract away the API and its potential changes? If it is then users of Dataplane shouldn't access original response fields no?

Reply:
> I'm not sure how I feel about the GeneratedDataplane being accessible on the end object.

Is this a general opinion? Or is that concern mainly related to creating the "UI type" (Dataplane here) using that type?

The value of having the generated/API type available in the project is that we can build the transformation from API type to UI type using a fully typed interface for which we would need to access that type.

Reply:
I think I've understood what your thinking is, but I may have misunderstood, to try to help:

I think I'd describe what I am trying to do here is to remove the sharp edges that exist when working with the backend responses in a frontend project. I don't think "fixing things" is the quite the right phrase because obviously nothing is "broken", but it "fixes" things that are difficult to work with in a frontend application. We do these "fixes" this in its own layer these so they are completely hidden from the application, the "fixes" could have very have happened in the backend. In fact I kinda pretend that they have.

The "fixes" are generally do 2 things that make these responses much easier to work with in the frontend:

1. Make sure no arrays or objects are `undefined` or `null`. Thus removing the majority of the cases for needing things like `??` in many places in the application and avoiding a whole class or errors.
2. Remapping commonly used data (sometimes aggregated or reshaped to make it more frontend friendly) into an easier place to access it, thus making other things in the application much much easier to build from scratch.

I guess the fact that I'm just smoothing off rough edges means I want to augment the existing generated type rather than having an entirely new type. I don't want a replacement as such, just tweak/smooth the existing ones. The original generated types should be hidden from the rest of the application and we pretend that our augmented types are the API types.

---

> If it is then users of Dataplane shouldn't access original response fields no?

As long as the original response fields need no "smoothing" (probably a better word), then I can't think of a reason why using those types to access the data wouldn't be ok.

We do copy/splat `generated` down onto a brand new object so JS-wise we aren't accessing the original object, TS-wise we are using the types from the original object but we probably want to make sure we do that.

Does that help at all?